src/types.jl

```diff
@@ -97,13 +52,416 @@ for Model in [:NeuralNetworkClassifier, :ImageClassifier]
+We start by defining a suitable `builder` object. This is a recipe
+for building the neural network. Our builder will work for images of
+any (constant) size, whether they be color or black and white (ie,
+single or multi-channel).  The architecture always consists of six
+alternating convolution and max-pool layers, and a final dense
+layer; the filter size and the number of channels after each
+convolution layer is customisable.
+```julia
+import MLJFlux
+
+struct MyConvBuilder
+    filter_size::Int
+    channels1::Int
+    channels2::Int
+    channels3::Int
+end
+
+make2d(x::AbstractArray) = reshape(x, :, size(x)[end])
+
+function MLJFlux.build(b::MyConvBuilder, rng, n_in, n_out, n_channels)
+    k, c1, c2, c3 = b.filter_size, b.channels1, b.channels2, b.channels3
+    mod(k, 2) == 1 || error("`filter_size` must be odd. ")
+    p = div(k - 1, 2) # padding to preserve image size
+    init = Flux.glorot_uniform(rng)
+    front = Chain(
+        Conv((k, k), n_channels => c1, pad=(p, p), relu, init=init),
+        MaxPool((2, 2)),
+        Conv((k, k), c1 => c2, pad=(p, p), relu, init=init),
+        MaxPool((2, 2)),
+        Conv((k, k), c2 => c3, pad=(p, p), relu, init=init),
+        MaxPool((2 ,2)),
+        make2d)
+    d = Flux.outputsize(front, (n_in..., n_channels, 1)) |> first
+    return Chain(front, Dense(d, n_out, init=init))
+end
+```
+It is important to note that in our `build` function, there is no final softmax. This is applie by default in all MLJFlux classifiers, using the `finaliser` hyperparameter of the classifier. Now that we have our builder defined, we can define the actual moel. If you have a GPU, you can substitute in `acceleration=CudaLibs()` below. Note that in the case of convolutions, this will **greatly** increase the speed of training.
+```julia
+ImageClassifier = @load ImageClassifier
+clf = ImageClassifier(builder=MyConvBuilder(3, 16, 32, 32),
+                      batch_size=50,
+                      epochs=10,
+                      rng=123)
+```
+You can add flux options such as `optimiser` and `loss` in the snippet above. Currently, `loss` must be a flux-compatible loss, and not an MLJ measure.
+Next, we can bind the model with the data in a machine, and fit the first 500 or so images:
+```julia
+mach = machine(clf, images, labels);
+
+fit!(mach, rows=1:500, verbosity=2);
+
+report(mach)
+
+chain = fitted_params(mach)
+
+Flux.params(chain)[2]
+```
+We can tack on 20 more epochs by modifying the `epochs` field, and iteratively fit some more:
+```julia
+clf.epochs = clf.epochs + 20
+fit!(mach, rows=1:500);
+```
+We can also make predictions and calculate an out-of-sample loss estimate, in two ways!
+```julia
+predicted_labels = predict(mach, rows=501:1000);
+cross_entropy(predicted_labels, labels[501:1000]) |> mean
+# alternative one liner!
+evaluate!(mach,
+          resampling=Holdout(fraction_train=0.5),
+          measure=cross_entropy,
+          rows=1:1000,
+          verbosity=0)
+```
+
```

Comment:
For the sake of keeping this concise, I suggest we end the example here and instead make some comments about the existence of general MLJ tools for controlling early stopping, etc .  We could add a link to `/examples/mnist`.